flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1505 @@
+		return requestCount;
+	}
+
+	@Override
+	public S get(K key, N namespace) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		long node = getNode(keyByteBuffer, 0, keyLen);
+
+		if (node == NIL_NODE) {
+			return null;
+		}
+
+		return helpGetNodeState(node);
+	}
+
+	@Override
+	public boolean containsKey(K key, N namespace) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		long node = getNode(keyByteBuffer, 0, keyLen);
+
+		return node != NIL_NODE;
+	}
+
+	@Override
+	public void put(K key, N namespace, S state) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+		byte[] value = skipListValueSerializer.serialize(state);
+
+		putNode(keyByteBuffer, 0, keyLen, value, false);
+	}
+
+	@Override
+	public S putAndGetOld(K key, N namespace, S state) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+		byte[] value = skipListValueSerializer.serialize(state);
+
+		return putNode(keyByteBuffer, 0, keyLen, value, true);
+	}
+
+	@Override
+	public void remove(K key, N namespace) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		removeNode(keyByteBuffer, 0, keyLen, false);
+	}
+
+	@Override
+	public S removeAndGetOld(K key, N namespace) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		return removeNode(keyByteBuffer, 0, keyLen, true);
+	}
+
+	@Override
+	public <T> void transform(
+		K key,
+		N namespace,
+		T value,
+		StateTransformationFunction<S, T> transformation) throws Exception {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		long node = getNode(keyByteBuffer, 0, keyLen);
+		S oldState = node == NIL_NODE ? null : helpGetNodeState(node);
+		S newState = transformation.apply(oldState, value);
+		byte[] stateBytes = skipListValueSerializer.serialize(newState);
+		putNode(keyByteBuffer, 0, keyLen, stateBytes, false);
+	}
+
+	// Detail implementation methods ---------------------------------------------------------------
+
+	/**
+	 * Find the node containing the given key.
+	 *
+	 * @param keyByteBuffer byte buffer storing the key.
+	 * @param keyOffset     offset of the key.
+	 * @param keyLen        length of the key.
```

Comment:
I would refrain from aligning the descriptions of every parameter description because it adds more maintenance burden in the long run. When seeing this, people have the urge to keep it like this. When renaming parameters which can cause this to break they need to touch the formatting of the JavaDoc as well.

Reply:
I thought this to be our code convention according to `StateMap`, maybe worth some discussion on ML? Thanks.